dataframe_show_reader/dataframe_assert_equal.py

```diff
@@ -0,0 +1,66 @@
+# Copyright 2019 The DataFrame Show Reader Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from pyspark.sql import DataFrame
+from pyspark.sql.types import StructField, StructType
+
+
+def assert_equal(expected: DataFrame, actual: DataFrame,
+                 show_output_on_failure: bool = True):
+    """Assert that two DataFrames contain the same data.
+    :param expected: The expected DataFrame
+    :param actual: The actual DataFrame
+    :param show_output_on_failure: If the DataFrames are not equal, show
+    the DataFrame schema or data (depending on where the mismatch is) in order
+    to help debugging.
+    :return: None
+    """
+    if expected is None:
+        assert actual is None, \
+            'The expected DataFrame is None, but the actual DataFrame is not.'
+        return  # Both DataFrames are None.
+    else:
+        assert actual is not None, \
+            'The actual DataFrame is None, but the expected DataFrame is not.'
+
+    expected.persist()
+    actual.persist()
+
+    # When comparing the schemas, ignore the metadata, which can include
+    # comments.
+    actual_schema_without_metadata = StructType(
```

Comment:
Yes, that is a good point. I added code to remove metadata from the expected schema as well. I also added the optional param `ignore_schema_metadata: bool = True` in case someone does actually have a desire to check that the metadata matches.

The order of the columns does matter. I have added this test to confirm and document that fact:
```
def test_assert_equal_when_column_order_is_different(
        expected_df, spark_session: SparkSession):
    actual_df = show_output_to_df("""
    +-----+-----+
    |col_b|col_a|
    +-----+-----+
    |1b   |1a   |
    |2b   |2a   |
    +-----+-----+
    """, spark_session)

    with raises(AssertionError) as exception_info:
        assert_equal(expected_df, actual_df)
    assert 'The DataFrame schemas differ.' == str(exception_info.value)
```